### ISSUE TITLE
Add an async interface

### ISSUE THREAD

**Issue by Marwes** [kkawakam/rustyline#126]
Probably needs to spawn an OS thread and perform all line reading there with a channel (I don't think there is another way to do it anyway). I could (an likely will) implement this outside `rustyline` but if it seems useful enough to warrant inclusion I am happy to do a PR.

**Comment by gwenn** [comment]
https://github.com/dpc/async-readline/blob/master/src/lib.rs#L165

**Comment by rubdos** [comment]
I was thinking about tokio support, which basically means async support. I'm probably also trying to sketch something up.

**Comment by gwenn** [comment]
https://docs.rs/tokio/0.1/tokio/io/struct.Stdin.html
> The handle implements the AsyncRead trait

**Comment by gwenn** [comment]
@RadicalZephyr I don't know how to make sure `rustyline` and the other/main thread update the screen correctly ?

**Comment by Ralith** [comment]
Very excited to see this receiving attention!

> I don't think Rustyline internals need to change to read from stdin asynchronously, I think it just needs to provide an async-compatible interface, which should probably be a futures [Stream](https://docs.rs/futures/0.1.25/futures/stream/trait.Stream.html) of `String`.

One reason to build async support into rustyline would be to allow operation without threads in cases where that is possible (e.g. stdin/out connected to a terminal or pipe on *nix). That's a lot of complexity for dubious benefit, though.

> I think for async use to be really nice, it would be good to have a new interface that allows output to be sent while the user is editing a line. I'm thinking of being able to create a remote "handle" to an `Editor` (kind of like the [`Handle`](https://docs.rs/tokio-core/0.1.17/tokio_core/reactor/struct.Handle.html) to a `Reactor` in the previous iteration of tokio. That handle would probably implement [`AsyncWrite`](https://docs.rs/tokio-io/0.1.11/tokio_io/trait.AsyncWrite.html).

Something like this sounds really great! Producing output without interfering with a prompt is exactly what I want for building e.g. an admin interface to a server. You might be able to get away with a synchronous `std::io::Write` interface, since nobody should be using this to dump large volumes of data anyway. Support for more fine-grained writes than complete lines will be useful for things like displaying live progress bars concurrent with user input.

**Comment by remexre** [comment]
It'd also be nice to be able to use async for output from the log crate; a stretch-goal-feature for this could be a wrapper for Handle that impls [`Log`](https://docs.rs/log/0.4.6/log/trait.Log.html).

**Comment by zyansheep** [comment]
If anyone still wants to add this feature to rustyline, I've made a cleanroom implementation that could be used as a reference.
https://github.com/zyansheep/rustyline-async

**Comment by esotericbyte** [comment]
I think this well intended but it's not necessarily going to achieve what you want it to. 
I would focus on what the goal is. I think aiming at first to interoperate with an application that is using asyc and tokio is a good goal. 
As it turns out threads are always available, and the threaded model must have a way of yielding time so that it can be put in a suspend state. Then the core can activate another thread. They don't all need to be hot. 
Also anything that uses async with tokio must be thread safe because tokio uses threads to spread tasks across cores. 

The core of a tokio application is usually an event loop that uses the tokio "select!" macro. 
This chooses one from among awaited futures that are in a ready state at random. 
Events filter into the select from among other things an mpsc channel. What is sent on a channel has to be pin and send which is complex way to say it needs to be in a fixed place in memory and usually fairly simple. We don't want to create a congestion point where the buffer being edited is sent on an mpsc channel so it should probably instead be shared with this main loop as an arc<mutex<buffer>>.  
Buffer would be whatever string type or data structure you use to manage what is being edited that hopefully has string_from or something like that implemented.  
A simple enum event type for when the buffer has changed, or the line is done can then be sent or returned to the application select! macro.  After it is handled a call back to rustyline would be a future awaited until another one of those events is available. while it's in an unready state it would not be selected and the application would be free to do the other things it does like network  IO  or handling other fields that are not line edited. 
I guess another less complex issue is that rustyline might be in different contexts within a complex app and so in each context it would have a different history and shared buffer. So maybe it's the app that passes this arc<mutex<buffer>> to rustyline and buffer contains both the line and history.  Then you 'd get back a Vec<Arc<mutex<buffer>> to save and load the histories and pass an unit16 to the select to indicate what context is being updated. 
I will look at the code and see if this makes sense.

**Comment by gwenn** [comment]
https://github.com/antirez/linenoise#asyncrhronous-api

**Comment by azoyan** [comment]
Are you planning to add this in the future?

**Comment by gwenn** [comment]
> Are you planning to add this in the future?

I am not sure but removing / replacing [ExternalPrinter](https://docs.rs/rustyline/latest/rustyline/trait.ExternalPrinter.html) by the clever / simple [linenoise](https://github.com/antirez/linenoise#asyncrhronous-api) solution is appealing

**Comment by dmlary** [comment]
> I am not sure but removing / replacing [ExternalPrinter](https://docs.rs/rustyline/latest/rustyline/trait.ExternalPrinter.html) by the clever / simple [linenoise](https://github.com/antirez/linenoise#asyncrhronous-api) solution is appealing.

I spent a few hours looking into this, and implementing an incremental read API similar to linenoise would be a significant architectural shift.

This isn't simply a matter of moving the needed state into the right place (currently the state is spread out across `Editor`, `State`, `InputState`, `impl Renderer`, and `impl RawReader`).  There's also a significant amount of recursion to handle command completion, reverse history search, paging, etc.  Those recursive routines would need to be rewritten to handle single `Cmd`s, and additional state added to determine which cmd handler function should we dispatch the next `Cmd` to when called.

Basically we'd need to keep a stack of `trait CmdDispatcher` instances, each holding whatever state they need, then passing it the result of `next_cmd()` each time readline updates.  There's a lot to figure out here, but there's a very high chance of breaking some existing behavior not covered by tests.

I think this approach is the right one as it would make it trivial to use rustyline in async, multi-threaded, and single-threaded multiplex (`select()`) situations.  @gwenn I'm just not sure you're interested in that significant of a change.  It would be a large PR to review.
